File: edrp-Backend/routers/notifications.py
from typing import List
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from auth import get_db, get_current_user
from models import User, Notification
from schemas import NotificationOut

router = APIRouter(prefix="/notifications", tags=["Notifications"])


@router.get("", response_model=List[NotificationOut])
def list_my_notifications(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return (
        db.query(Notification)
        .filter(Notification.user_id == current_user.id)
        .order_by(Notification.created_at.desc())
        .limit(50)
        .all()
    )


@router.get("/unread-count")
def get_unread_count(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    count = (
        db.query(Notification)
        .filter(Notification.user_id == current_user.id, Notification.is_read == False)
        .count()
    )
    return {"unread_count": count}


@router.patch("/{notification_id}/read", status_code=204)
def mark_as_read(
    notification_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    note = db.query(Notification).filter(Notification.id == notification_id).first()
    if not note:
        raise HTTPException(status_code=404, detail="Notification not found")
    if note.user_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not your notification")

    note.is_read = True
    db.commit()
    return None


@router.patch("/mark-all-read", status_code=204)
def mark_all_read(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    db.query(Notification).filter(
        Notification.user_id == current_user.id, Notification.is_read == False
    ).update({Notification.is_read: True})
    db.commit()
    return None

File: frontend/src/api/notifications.ts
import client from "./client";

export interface NotificationItem {
  id: number;
  user_id: number;
  title: string;
  message: string;
  type: string;
  link_url?: string;
  is_read: boolean;
  created_at: string;
}

export const notificationsApi = {
  getNotifications: async (unreadOnly = false, limit = 50) => {
    const response = await client.get<NotificationItem[]>("/api/notifications/", {
      params: { unread_only: unreadOnly, limit },
    });
    return response.data;
  },

  getUnreadCount: async () => {
    const response = await client.get<{ unread_count: number }>("/api/notifications/unread-count");
    return response.data.unread_count;
  },

  markAsRead: async (id: number) => {
    const response = await client.patch<NotificationItem>(`/api/notifications/${id}/read`);
    return response.data;
  },

  markAllAsRead: async () => {
    const response = await client.patch<{ message: string; updated: number }>("/api/notifications/read-all");
    return response.data;
  },
};
